### ISSUE TITLE
Retry not working correctly for large files

### ISSUE THREAD

**Issue by Fra-nk** [Azure/azure-storage-python#449]
### Which service(blob, file, queue) does this issue concern?
Blob

### What problem was encountered?
Upload of a large file had a failing blob block upload.
Azure storage says it correctly re-uploaded, but according to Storage analytics that did not happen:

* `<Agent>=Azure-Storage/1.1.0-1.1.0 (Python CPython 2.7.9; Linux 3.16.0-5-amd64)`
* `MAX_BLOCK_SIZE=100000000` = 100 MB
* `max_connections=8`
* `timeout=30`
* Total Blob size: `4294903296` = 4 GiB
* Only `4249585152` were uploaded (checked with `az storage blob show`)

Azure Storage of the particular block
```
2018-04-18 16:33:04,807 <az> INFO  Client-Request-ID=2b3d9568-4326-11e8-8ef0-000d3a29b453 Outgoing request: Method=PUT, Path=/<container>/<blob>?<SAS>, Query={'comp': 'block', 'blockid': u'QmxvY2tJZDAwMDAx', 'timeout': '30'}, Headers={'Content-Length': '100000000', 'x-ms-client-request-id': '2b3d9568-4326-11e8-8ef0-000d3a29b453', 'User-Agent': '<Agent>', 'x-ms-version': '2017-07-29', 'x-ms-lease-id': None, 'x-ms-date': 'Wed, 18 Apr 2018 16:33:04 GMT'}.
2018-04-18 16:33:38,838 <az> INFO  Client-Request-ID=2b3d9568-4326-11e8-8ef0-000d3a29b453 Operation failed: checking if the operation should be retried. Current retry count=0, , HTTP status code=Unknown, Exception=SSLError: HTTPSConnectionPool(host='<account>.blob.core.windows.net', port=443): Max retries exceeded with url: /<container>/<blob>?<SAS>&comp=block&blockid=QmxvY2tJZDAwMDAx&timeout=30 (Caused by SSLError(SSLError('The write operation timed out',),)).
2018-04-18 16:33:53,265 <az> INFO  Client-Request-ID=2b3d9568-4326-11e8-8ef0-000d3a29b453 Outgoing request: Method=PUT, Path=/<container>/<blob>?<SAS>, Query={'comp': 'block', 'blockid': u'QmxvY2tJZDAwMDAx', 'timeout': '30'}, Headers={'Content-Length': '100000000', 'x-ms-client-request-id': '2b3d9568-4326-11e8-8ef0-000d3a29b453', 'User-Agent': '<Agent>', 'x-ms-version': '2017-07-29', 'x-ms-lease-id': None, 'x-ms-date': 'Wed, 18 Apr 2018 16:33:53 GMT'}.
2018-04-18 16:33:59,569 <az> INFO  Client-Request-ID=2b3d9568-4326-11e8-8ef0-000d3a29b453 Receiving Response: Server-Timestamp=Wed, 18 Apr 2018 16:33:58 GMT, Server-Request-ID=2c9b33ae-701e-0109-6133-d7286a000000, HTTP Status Code=201, Message=Created, Headers={'Content-Length': '100000000', 'x-ms-client-request-id': '2b3d9568-4326-11e8-8ef0-000d3a29b453', 'User-Agent': '<Agent>', 'x-ms-version': '2017-07-29', 'x-ms-lease-id': None, 'x-ms-date': 'Wed, 18 Apr 2018 16:33:53 GMT'}.
```
Storage Analytics of the particular block:
```
1.0;2018-04-18T16:33:53.3070387Z;PutBlock;SASSuccess;201;6260;5669;sas;;<account>;blob;"https://<account>.blob.core.windows.net:443/<container>/<blob>?<SAS>&amp;comp=block&amp;blockid=QmxvY2tJZDAwMDAx&amp;timeout=30";"/<account>/<container>/<blob>";2c9b33ae-701e-0109-6133-d7286a000000;0;<internal IP address>:53156;2017-07-29;567;54681856;193;0;54681856;;"zNZ1j2PNDLDEV5szpt0DKg==";;;;"<Agent>";;"2b3d9568-4326-11e8-8ef0-000d3a29b453"
```

* Azure Storage sees `'Content-Length': '100000000'`, while Storage Analytics **says** it only received `54681856` bytes.
    * `100000000 - 54681856 = 45318144`
    * `4294903296 - 4249585152 = 45318144`
* Storage Analytics only registers the black after the retry started.
* Storage Analytics did show no data for the allegedly successful upload.

I can also provide you with the other logs of the successfully uploaded blocks, etc., but they contain little information:
Content-length is consistent, return-codes are ok (201), timestamps are reasonable, blocklist/head upload happens in the correct order, etc.

I can not reliably reproduce the error, but I see it occasionally when uploading several TiB from on-premise as well as Azure-VMs.

### Have you found a mitigation/solution?
Not yet, but complete re-upload of the block *after* the blocklist has been committed seems sensible.

**Comment by zezha-msft** [comment]
Hi @Fra-nk, thank you for reaching out! And I apologize for the inconvenience.

The issue seems to be caused by a bug that I already [fixed](https://github.com/Azure/azure-storage-python/commit/f77b143815e95ad05444718b9316972ba58e97bd) but have not yet published. It will go out with the next release very soon.

Btw, are you using the `create_blob_from_path` method? Or are you calling `put_block` and `put_block_list`?

**Comment by Fra-nk** [comment]
Hi @zezha-msft,
thanks for the quick answer and for your work on this.
I'll try to give the `dev` branch a try, then 😃.

In my case, I'm calling `create_blob_from_path`.
I assume using a lower level implementation and explicitly `try`ing `put_block` would solve the problem, but [in my use case](https://github.com/henrysher/duplicity/blob/master/duplicity/backends/azurebackend.py#L113) that's only reasonable for testing, since I have to modify upstream code, which aims to use the high-level implementation.

**Comment by zezha-msft** [comment]
Hi @Fra-nk, calling `create_blob_from_path` is perfectly fine, I was just making sure that the issue occurred when a seekable stream was used as body. Basically the bug was that we weren't rewinding the body (seekable stream) properly in the case of a retry. And with larger files, it is more likely that we have retries occurring and thus encountering this bug. It should now work properly after the [fix](https://github.com/Azure/azure-storage-python/commit/f77b143815e95ad05444718b9316972ba58e97bd) on dev branch. 

Please let me know if you encounter any other problem or have any question. I'll keep this issue open until the fix gets released. Thank you!

**Comment by Fra-nk** [comment]
Hi,
positive feedback from my side:
I ran tests uploading ~20 TiB to a block blob using the latest `dev` branch and the upload went smoothly: Some retries occurred but `azure-storage` handled them correctly.
Good job with the fix and thanks a lot 😃 for your work.
From my side the issue itself can be considered resolved, but as you said, it's probably best to keep it open until the fix gets released for visibility.

Can you give a _rough estimate_ for
> It will go out with the next release **very soon**.

Are we talking a few days or a few weeks?

**Comment by zezha-msft** [comment]
Hi @Fra-nk, I'm glad to hear that!

The release is going out in 1-2 weeks, as we have a few more changes to merge in. Thank you!

**Comment by zezha-msft** [comment]
The fix is in 1.2.0rc1. I'll close this issue once the preview period of 1.2.0 is over.

**Comment by zezha-msft** [comment]
Published in version 1.3.0.
